pkg/auth/cred.go

```diff
@@ -30,26 +30,19 @@ func NewCredential(cfg *Config) (azcore.TokenCredential, error) {
 		return nil, fmt.Errorf("failed to create credential, nil config provided")
 	}
 
-	if cfg.UseCredentialFromEnvironment {
+	if cfg.ArmAuthMethod == authMethodWorkloadIdentity {
 		klog.V(2).Infoln("cred: using workload identity for new credential")
 		return azidentity.NewDefaultAzureCredential(nil)
```

Comment:
"using workload identity" - this is not quite what `NewDefaultAzureCredential` does, it will [attempt a variaty of mechanisms](https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/azidentity/README.md#defaultazurecredential). If we are Ok with this - which may well be the case - let's just say it, and we don't need to special case Managed Identity below. If we want to constrain this to Workload Identity - needs to use `NewWorkloadIdentityCredential`?

Reply:
There is a discussion related to this with @charliedmcb in another PR. I think our understanding match yours, with the new knowledge here is that `NewDefaultAzureCredential()` might cover managed identity as well.

So far, I think that we should scope it down to workload identity for simplicity. This means we will support only two auth methods for now: workload identity and system assigned MSI.
In the current implementation, there might be "shortcuts" where `ARM_AUTH_METHOD` is workload identity, but they ended up using something else that's compatible with this implementation. 

So, I think I will make changes to `NewWorkloadIdentityCredential`, if possible.

Reply:
I'm not sure what you mean by below; could you provide more example? 
>  Would be good to avoid custom switches (like ARM_AUTH_METHOD) if the "standard"/OOB works well.

Reply:
What I mean is that `azidentity` already provides (via environment variables) an existing configuration surface, and (via `DefaultAzureCredential`) a reasonable sequence of auth methods to try, including the two that we need to support. We should have good reasons to a) constrain the auth methods, and b) introduce additional/custom configuration surface, such as the proposed `ARM_AUTH_METHOD` setting. (Which we may need to remove later if decide to use `NewDefaultAzureCredential`? Removal is always hard ...) There may be valid reasons for both (improved security posture, the built-in sequence not meeting our order requirements, established pattern elsewhere, etc.) - but would be good to have clarity on which ones - if any - apply here.

Reply:
One more option to consider - if we decide there is a good reason to support only workload identity and MSI - is to just try them in order (similar to how `DefaultAzureCredential` does it), eliminating the need (and capability) for the explicit choice via `ARM_AUTH_METHOD`.

Reply:
>  b) introduce additional/custom configuration surface, such as the proposed ARM_AUTH_METHOD setting

`ARM_AUTH_METHOD` is in fact a replacement for the awkward combination of `ARM_USE_CREDENTIAL_FROM_ENVIRONMENT` and `ARM_USE_MANAGED_IDENTITY_EXTENSION` rather than an addition. 
But, if we use `DefaultAzureCredential`, maybe that can be implied from the credentials and the user don't have to explicitly specify `ARM_AUTH_METHOD`, eliminating its needs.

Reply:
> a) constrain the auth methods

I find that we never(?) have an expectation of the user using other auth methods than these two in the first place. Even if we allow it, that allow means additional responsibility/accountability (outside of this relatively simple implementation, I mean) on our side. 
So, unless there are good demands for other auth methods that is possible through `DefaultAzureCredential`, I don't think we should give the opportunity for the user to mess up things that we not yet want to invest into.

But at least, our code framework should be able to support that quickly once we decided to, but not exposing it to the user.

Reply:
> Removal is always hard ...
Removing ARM_AUTH_METHOD to eliminate user-side complexity, given `DefaultAzureCredential` can retrieve that implicitly, is an annoyance on our side to change the code, but I think it is ok.

On the other hand, if the user secretly rely on some auth methods that `DefaultAzureCredential` supports, but then we decide to remove it (e.g., security), then that's going to affect them.
Technically, if we stop using it and switch to `NewWorkloadIdentityCredential`, it is like that. But given we are in the early stage, I think it is better to do it now.

Reply:
Keep in mind that whatever in `DefaultAzureCredential` today might changes in the future under our radar, given its definition is open to new methods rather than always specific to a few already.

Reply:
My preference would be for the simplest secure approach that supports the requirements, which seems to be `DefaultAzureCredential`. (Any changes there will likely require additional configuration, and are very unlikely to break existing auth scenarios. In NAP we control the configuration tightly enough for this not to be a concern, for self-hosted breaking changes, if any, would/should surface via release notes.) Second best would be to try a more controlled sequence of auth methods we need to support - unless you see a problem with this, and we have to specify the method explicitly via configuration.

Reply:
Agreed DefaultAzureCredential already encapsulates what we need. I don't see them breaking it either as its wildly used in many projects for wrapping many types of auth based on some quick github code searching.

Reply:
> Any changes there will likely require additional configuration, and are very unlikely to break existing auth scenarios

> I don't see them breaking it either

I doubt it will break too, plus the tests will catch it anyway in case we neglect release notes. 
But my point is it might have new auth methods we are not aware of, which means more liability as I mentioned. This will not happen in the more controlled approach (i.e., `NewWorkloadIdentityCredential` with `ARM_AUTH_METHOD`)

>  In NAP we control the configuration tightly enough for this not to be a concern

Yep all these concerns are for self-hosted. In NAP we have static integration interface.

Reply:
But maybe I am overthinking again. We could try going with implicit auth method and use `DefaultAzureCredential` and see how it goes.

The only problem, if it is at all, is `autorest_auth.go` is also using it. But I just haven't take a look at that deeply yet. If we are moving in this direction then I will see the possibility.

Reply:
Circling back to this. 
- The concern of unwanted auth method is valid, from an example in Cluster Autoscaler. "Requirement" to continue the support is an engineering burden.
- Although, this `DefaultAzureCredential` is something standardized. If we are going to rely on this standardization (cloud-provider-azure?) for a next while, the simplicity we get might worth it, and the deprecation (if necessary) might not be too difficult.

TL;DR let's just use `DefaultAzureCredential`.